File: app/src/main/java/mk/webfactory/template/log/CrashlyticsLogger.java
package mk.webfactory.template.log;

import android.content.Context;
import android.os.Build;
import android.util.DisplayMetrics;
import android.util.Log;
import com.crashlytics.android.Crashlytics;
import com.crashlytics.android.core.CrashlyticsCore;
import java.util.ArrayList;
import java.util.List;
import timber.log.Timber;

public final class CrashlyticsLogger extends Timber.Tree implements CrashReportLogger {

  private static final int PAGES_LIST_SIZE = 5;

  private final Context context;
  private final CrashlyticsCore crashlytics;
  private final CrashlyticsExceptionHandler exceptionHandler;
  private final ArrayList<String> lastUsedPages;

  public CrashlyticsLogger(Context context, Crashlytics crashlytics) {
    this.crashlytics = crashlytics.core;
    this.context = context;
    this.lastUsedPages = new ArrayList<>(PAGES_LIST_SIZE);
    this.exceptionHandler = new CrashlyticsExceptionHandler(crashlytics.core, context);
    logDeviceInfo();
  }

  private void logDeviceInfo() {
    final DisplayMetrics dm = context.getResources().getDisplayMetrics();
    crashlytics.setString("screenResolution", dm.heightPixels + "x" + dm.widthPixels);
    crashlytics.setFloat("screenDensity", dm.density);
    crashlytics.setInt("apiLevel", Build.VERSION.SDK_INT);
    crashlytics.setString("manufacturer", Build.MANUFACTURER);
    crashlytics.setString("model", Build.MODEL);
  }

  @Override public void setCurrentPage(String page) {
    crashlytics.setString("currentPage", page);
    lastUsedPages.add(page);
    if (lastUsedPages.size() > PAGES_LIST_SIZE) {
      lastUsedPages.remove(0);
    }
  }

  @Override public List<String> getLastUsedPages() {
    return lastUsedPages;
  }

  @Override protected void log(int priority, String tag, String message, Throwable t) {
    if (priority == Log.VERBOSE || priority == Log.DEBUG) {
      return;
    }
    crashlytics.log(priority, tag, message);
    if (t != null) {
      crashlytics.logException(t);
    }
  }
}

File: app/src/main/java/mk/webfactory/template/log/CrashlyticsLogger.kt
package mk.webfactory.template.log

import android.content.Context
import android.os.Build
import android.util.Log
import com.google.firebase.crashlytics.FirebaseCrashlytics
import timber.log.Timber
import java.util.*

class CrashlyticsLogger(
    private val context: Context,
    crashlytics: FirebaseCrashlytics
) :
    Timber.Tree(), CrashReportLogger {
    private val crashlytics: FirebaseCrashlytics = crashlytics
    private val exceptionHandler: CrashlyticsExceptionHandler
    private val lastUsedPages: ArrayList<String>

    init {
        lastUsedPages = ArrayList(PAGES_LIST_SIZE)
        exceptionHandler = CrashlyticsExceptionHandler(crashlytics, context)
        logDeviceInfo()
    }

    private fun logDeviceInfo() {
        val dm = context.resources.displayMetrics
        crashlytics.setCustomKey("screenResolution", dm.heightPixels.toString() + "x" + dm.widthPixels)
        crashlytics.setCustomKey("screenDensity", dm.density)
        crashlytics.setCustomKey("apiLevel", Build.VERSION.SDK_INT)
        crashlytics.setCustomKey("manufacturer", Build.MANUFACTURER)
        crashlytics.setCustomKey("model", Build.MODEL)
    }

    override fun setLoggedInUser(userIdentifier: String) {
        crashlytics.setUserId(userIdentifier)
    }

    override fun setCurrentPage(page: String) {
        crashlytics.setCustomKey("currentPage", page)
        lastUsedPages.add(page)
        if (lastUsedPages.size > PAGES_LIST_SIZE) {
            lastUsedPages.removeAt(0)
        }
    }

    override fun lastUsedPages(): List<String> {
        return lastUsedPages
    }

    override fun log(priority: Int, tag: String?, message: String, t: Throwable?) {
        if (priority == Log.VERBOSE || priority == Log.DEBUG) {
            return
        }
        crashlytics.log(message)
        if (t != null) {
            crashlytics.recordException(t)
        }
    }

    companion object {
        private const val PAGES_LIST_SIZE = 5
    }
}